OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs

```diff
@@ -65,6 +73,175 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
 
 			hueSlider.Value = initialColor.H / 255f;
 			onChange(mixer.Color);
+
+			// Setup tab controls
+			var mixerTab = widget.GetOrNull<BackgroundWidget>("MIXER_TAB");
+			var paletteTab = widget.GetOrNull<BackgroundWidget>("PALETTE_TAB");
+			var paletteTabScrollPanel = widget.GetOrNull<ContainerWidget>("PALETTE_TAB_SCROLLPANEL");
+			var mixerTabButton = widget.GetOrNull<ButtonWidget>("MIXER_TAB_BUTTON");
+			var paletteTabButton = widget.GetOrNull<ButtonWidget>("PALETTE_TAB_BUTTON");
+			var presetArea = paletteTabScrollPanel.GetOrNull<ContainerWidget>("PRESET_AREA");
+			var customArea = paletteTabScrollPanel.GetOrNull<ContainerWidget>("CUSTOM_AREA");
+			var presetColorTemplate = paletteTabScrollPanel.GetOrNull<ColorBlockWidget>("COLORPRESET");
+			var customColorTemplate = paletteTabScrollPanel.GetOrNull<ColorBlockWidget>("COLORCUSTOM");
+
+			mixerTabButton.OnClick = () =>
+			{
+				mixerTab.Visible = true;
+				paletteTab.Visible = false;
+				paletteTabOpenedLast = false;
+			};
+			paletteTabButton.OnClick = () =>
+			{
+				paletteTab.Visible = true;
+				mixerTab.Visible = false;
+				paletteTabOpenedLast = true;
+			};
+			mixerTabButton.IsHighlighted = () => mixerTab.Visible;
+			paletteTabButton.IsHighlighted = () => paletteTab.Visible;
+			if (paletteTabOpenedLast)
+			{
+				paletteTab.Visible = true;
+				mixerTab.Visible = false;
+			}
+
+			// Setup the mod team preset colors
+			int maxCustomColors = 8;
+			int maxColumns = 8;
+			int swatchPadding = 0;
+
+			// Build the color swatch controls
+			var addSwatchesAction = new Action<string[], Widget, ColorBlockWidget, bool>((colorList, containerWidget, template, isEditable) =>
+			{
+				int columnIndex = 0;
+				string controlName = template.Id;
+				int x = template.Bounds.X;
+				int y = template.Bounds.Y;
+				int swatchSize = template.Bounds.Width;
+
+				int initX = x;
+
+				for (int i = 0; i < colorList.Length; i++)
+				{
+					var nameString = controlName + i;
+					var newSwatch = new ColorBlockWidget()
```

Comment:
This can simplify to `newSwatch = template.Clone()`

Reply:
Is it alright if it looks like this afterward?

```
var newSwatch = (ColorBlockWidget) template.Clone();
newSwatch.IsVisible = () => true;
newSwatch.Id = nameString;
newSwatch.X = x.ToString();
newSwatch.Y = y.ToString();
newSwatch.Initialize(new WidgetArgs());
```